### DISCUSSION TITLE
[Feature] Memories Widgets

### DISCUSSION THREAD

**Discussion by Brooklyn-Mark** [immich-app/immich#6082] [Feature Request]
### The feature

iOS widgets of different types of memories for the Home Screen or slide over area like Apple Photos or Google Photos does. Then if you click opens the Immich app and shows that memory.

### Platform

- [ ] Server
- [ ] Web
- [X] Mobile

**Comment by GioPan04** [comment]
This could be possible using [home_widget](https://pub.dev/packages/home_widget) and [workmanager](https://pub.dev/packages/workmanager) packages.

**Reply by mvamsee2010** [reply]
This would be a much wanted feature! Looking forward to it. Thanks.

**Reply by SerAlbi** [reply]
I'm so looking forward to this feature! Even better if it has some kind of 'smart stories' Ai feature like iPhone, for example recognizing your partner or a new trip!

**Reply by weiyideai520** [reply]
I also need it, and I hope the author will consider it a priority.

**Reply by mrreview** [reply]
I also hope to see this feature soon!

**Comment by NickDiMucci** [comment]
Widget for Android would also be great.

**Comment by owpac** [comment]
I would love to have widgets with my photos shown randomly!

**Comment by aijunji2009** [comment]
My primary way of enjoying photos is through widgets.
This is the only reason I am still hesitant to replace Apple Photos with Immich。

**Comment by quinnaissance** [comment]
Just a request on top of this, it would be cool if there was some sort of random seed determining the order of photos, so for shared albums, multiple people can use the same seed to sync their widgets together.

**Reply by dkerlee** [reply]
I have a lot of random photos. I would way rather have some sort of mix of faces, or people doing cool things, other than one of the lame pics I have!

**Comment by DonDestroY** [comment]
Just moved to Immich and this is Wahoo, I mean the completness and user experience is impressive ! Upvote on this feature that could be very usefull :-)

**Comment by dwnsdp** [comment]
Here to say yes please!

**Comment by Xtylish99** [comment]
I'd love to have widget on my phone

**Comment by tumstech** [comment]
Recently switched to Immich from Piwigo, and here to stay.
Widgets to both IOS and Android sound awesome. Waiting for them to be available in the future release.

**Comment by bo0tzz** [comment]
Folks, as is hopefully pretty clear from all the minimized comments, if you don't have anything to add to the request please just upvote the original post.

**Comment by quinnaissance** [comment]
So I had an idea that you could accomplish this on iOS with a [Scriptable](https://scriptable.app) widget that uses the Immich API, as a temporary workaround until it's implemented natively. I gave it a shot to see if it was feasible.

It's rather janky but works. Image quality is hit or miss. I have put up a [gist](https://gist.github.com/quinnaissance/4d7eeab657f2d0220bb34de754f1f59d) for this in case anyone wants to try it out.

**Comment by helblinglilly** [comment]
I didn't realise people used this for memories!

I personally like to select a photo album and then have it pick out random photos from there. Nice after just having been on a holiday, or picking out landscape albums with a certain vibe to them.

First time in this space - is this likely to be prioritised and done soon or would you be open to contributions on this?

**Reply by alextran1502** [reply]
@helblinglilly we don't have this in our lists at the moment, so contribution is welcomed

**Comment by gduhachek** [comment]
We really need this feature for immich to be a serious competitor. Even Synology Photos has a brilliant memories widget. I'd like to see this before many other features that are being worked on!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

**Comment by bwees** [comment]
We have merged in #19148 which adds support for iOS widgets and lays the groundwork for Android widgets to follow soon. Look for it in the upcoming v1.135.0 release!

**Reply by quinnaissance** [reply]
This is great work! Just out of curiosity, did you test it side by side with the iOS Photos widget? Sometimes the crops seem a little too perfect, so I've always wondered if it crops based on detected subjects, but I've never found any info online about this.

**Reply by bwees** [reply]
> This is great work! Just out of curiosity, did you test it side by side with the iOS Photos widget? Sometimes the crops seem a little too perfect, so I've always wondered if it crops based on detected subjects, but I've never found any info online about this.

iOS appears to crop based on faces if they are present. This is something that could be added in the future to Immich but is not present in the current implementation

**Reply by nordewal** [reply]
I rarely regretted my switch from iPhone->Android ~15y ago. But this is one such rare moment - enjoy it Apple folks 😀
Way to go @bwees!

**Reply by bwees** [reply]
> I rarely regretted my switch from iPhone->Android ~15y ago. But this is one such rare moment - enjoy it Apple folks 😀 Way to go @bwees!

Thanks! One of my next projects is to implement Android widgets so hopefully that comes soon :)

**Reply by SneakieGargamel** [reply]
This is very nice! Will there also be Person widget in the future? You select 1 or maybe multiple persons and get there images? Great feature that wasn’t even on the roadmap

**Reply by bwees** [reply]
> This is very nice! Will there also be Person widget in the future? You select 1 or maybe multiple persons and get there images? Great feature that wasn’t even on the roadmap

This is definitely something that could be added in the future but no plans at this time. Current focus is to get widgets on android then I can look at more widget options!

**Reply by SneakieGargamel** [reply]
> > This is very nice! Will there also be Person widget in the future? You select 1 or maybe multiple persons and get there images? Great feature that wasn’t even on the roadmap
> 
> This is definitely something that could be added in the future but no plans at this time. Current focus is to get widgets on android then I can look at more widget options!

No problem ofcourse, just wanted to pass along a users wish. Good luck with the Andriod version!

**Reply by Brooklyn-Mark** [reply]
Hey @bwees thanks for implementing this! I do have an issue and I'm not sure if this is known, there is a fix at all, or, it's just a current limitation to how this is coded.

See I use built-in tinting feature with iOS 18.5 to color my icons and having this enabled causes the widget to just display a white box. This happened with some other widgets too until they updated something so I know it's fixable but maybe not with photo based stuff.

Just something to be aware of. Cheers!

**Reply by bwees** [reply]
@Brooklyn-Mark thanks for the heads up. Doing some more improvements to the iOS widget today so ill add this to the list

**Reply by Brooklyn-Mark** [reply]
@bwees Awesome! appreciate it.

**Reply by bwees** [reply]
@Brooklyn-Mark keep an eye on #19893

**Reply by phil9309** [reply]
> 
> Thanks! One of my next projects is to implement Android widgets so hopefully that comes soon :)

Any news here? My wife is dying for this widget, it' her favorite feature by far!

**Reply by bwees** [reply]
> > Thanks! One of my next projects is to implement Android widgets so hopefully that comes soon :)
> 
> Any news here? My wife is dying for this widget, it' her favorite feature by far!

It is in the review pipeline at #19310

**Comment by bwees** [comment]
Android widget support has been merged with #19310. Keep an eye out in the next release!

**Comment by bverkron** [comment]
@bwees not sure if you would prefer a new Feature Request for this or not but a key feature missing from the iOS widget IMO is being able to jump to the exact photo in Immich when you tap on the widget. For example when you see a photo displayed in the widget and you tap the widget it should take you directly to that photo in Immich instead of just opening Immich and requiring you to search for that image, which can be challenging to find sometimes.

Maybe this capability already exists and I'm just missing something. Either way having the widget in general is HUGE and I love it!

**Reply by bwees** [reply]
This functionality is already merged, keep an eye out for it in the next release!

**Comment by nordewal** [comment]
w00t - just tried out the Android widget with the new release, this is awesome :heart: 

I might have found a bug: I am using the mTLS to connect to my immich server. When opening the settings for the random widget it complains about `No connection to the server is available. Please try again later` The immich app itself works flawlessly meanwhile. When I disable the mTLS requirement on my server, the widget works without an issue.

The memories widget seems to work with mTLS without an issue :)

**Reply by bwees** [reply]
mTLS among other HTTP features have known issues (see #15230). It is interesting that only the memories widget works but mTLS is extremely inconsistent as reported in that thread. From my understanding, the issue appears to come from internal system network APIs, and there is not much we can do until Android fixes their support.

**Comment by Brooklyn-Mark** [comment]
@bwees thanks for the fixes. Finally got some time to respond and ask for some improvements. Would it be possible to do what apple photos does on tinted backgrounds and make the photos black & white and not add the tint to the image? Looks quite good. Then they use a block white font and only tint the words so that they don’t need to use an overlay. This allows you to see more of the image. See my screenshot example (ignore the blurred faces)
![IMG_3158](https://github.com/user-attachments/assets/c849b5cf-cc2c-4ad7-bfac-42ad0a74b7d5)
Vs this:

**Reply by Brooklyn-Mark** [reply]
![IMG_3159](https://github.com/user-attachments/assets/a8a106b6-5f1d-4180-a393-60a57dff2537)

**Comment by mehalter** [comment]
I'm using the Android widget now and it is great! One thing that would be a great additional feature is to have the option to do a random image from a specific face rather than an album or memories. I currently am working around this by using the `immich-face-to-album` project to maintain an album of a collection of faces and using that with the widget, but it would be nice to have a native solution!

**Comment by jrasm91** [comment]
This has been implemented.

**Reply by DaJansenGit** [reply]
Do you refer to the face based memories suggested by mehalter? Can't find it in the widget options. Server and app on v. 2.1.0

**Reply by bwees** [reply]
He is referring to widgets as a whole. Face based memories and albums is a larger feature request outside of widgets. The memory feature is tracked in #2836.

If/when the feature is implemented, it will be documented in the release notes.

**Reply by DaJansenGit** [reply]
Ok, thanks! Used the `immich-face-to-album` repository as well, works like a charms with a docker container for now. Thanks for the awesome project.
